### PR TITLE
Order signature

### PR DESCRIPTION
**Pull request by sajberkurajber90**

### CODE REVIEW

**Review by SirSplasch — CHANGES_REQUESTED** [review]
Also I'm unsure whether the expected functionality is given. If I create the pdf file of the order the signature is not on it, even though I added one via the public link.

File: src/Livewire/PublicLinkGenerator.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    public function setPublicLink(): void
+    {
+        $this->publicLink = URL::signedRoute('order.public', ['order' => $this->order->uuid]);
+    }
+
+    #[Computed]
+    public function getSignature(): bool
+    {
+        return is_null(
```
Comment (patrickweh) [inline src/Livewire/PublicLinkGenerator.php:27]: not is_null better do just a ->exists() instead of ->first()

File: src/Livewire/PublicLinkGenerator.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    #[Modelable]
+    public OrderForm $order;
+
+    public function setPublicLink(): void
+    {
+        $this->publicLink = URL::signedRoute('order.public', ['order' => $this->order->uuid]);
+    }
+
+    #[Computed]
+    public function getSignature(): bool
```
Comment (patrickweh) [inline src/Livewire/PublicLinkGenerator.php:25]: rename to hasSignature

File: resources/views/livewire/public-link-generator.blade.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<div>
+    @if($this->getSignature && is_null($publicLink))
```
Comment (patrickweh) [inline resources/views/livewire/public-link-generator.blade.php:2]: rename to hasSignature

File: resources/views/livewire/public-link-generator.blade.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        <x-button
+            class="w-full"
+            wire:click="setPublicLink"
+            :label="__('Create Signature')"
```
Comment (patrickweh) [inline resources/views/livewire/public-link-generator.blade.php:6]: "Add signature" is maybe better than create

File: resources/views/livewire/order/public-link.blade.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                        </template>
+                        <template x-if="id && !error">
+                            <p class="text-2xl">{{__("Signature saved")}}</p>
+                        </template>
+                    </div>
+                </div>
+                <canvas x-ref="canvas" width=500 height=200 class="rounded-md h-auto">
+                </canvas>
+            </div>
+            <div class="mt-4 h-8 mb-4 flex justify-end w-[100%] space-x-5">
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:27]: why w-[100%]? isnt it the same as w-full?

File: resources/views/livewire/order/public-link.blade.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                                <x-icon name="exclamation"/>
+                            </template>
+                            <template x-if="id && !error">
+                                <x-icon name="check"/>
+                            </template>
+                        </div>
+                        <template x-if="error">
+                            <p class="text-2xl">{{__("Signature Upload Failed")}}</p>
+                        </template>
+                        <template x-if="id && !error">
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:19]: why not x-show?

File: resources/views/livewire/order/public-link.blade.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    <div x-data="signature($wire,$refs)" class="bg-gray-100 pt-10 pr-10">
+        <div class="flex flex-col items-end">
+            <div class="flex w-full justify-between">
+                <div class="flex-1 flex items-center justify-center">
+                    <div x-cloak class="flex flex-col items-center justify-center" x-show="error || id">
+                        <div class="w-10 h-10">
+                            <template x-if="error">
+                                <x-icon name="exclamation"/>
+                            </template>
+                            <template x-if="id && !error">
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:12]: why not x-show?

File: resources/views/layouts/empty.blade.php [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
```
Comment (patrickweh) [inline resources/views/layouts/empty.blade.php:1]: you dont need to include livewire styles and scripts in livewire v3.
also dont use the vite helper, better use:
```
    {{ Asset::toHtml(
            ViewServiceProvider::getRealPackageAssetPath(
                '/resources/js/alpine.js',
                'team-nifty-gmbh/flux-erp'
            )
        )
    }}
```

File: lang/de.json [hunk 9]
```diff
@@ -1056,6 +1057,8 @@
     "Show parent": "Übergeordneten anzeigen",
     "Show unassigned payments": "Nicht zugeordnete Zahlungen anzeigen",
     "Showing": "Zeige",
+    "Signature saved": "Unterschrift gespeichert",
+    "Signature Upload Failed": "Upload fehlgeschlagen",
```
Comment (patrickweh) [inline lang/de.json:1061]: Rename occurrences from Signature Upload Failed to Upload failed and remove this translation

**Review by patrickweh — CHANGES_REQUESTED** [review]

File: resources/js/components/signature-pad.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+                this.error = true;
+            }
+        },
+        async save() {
+            if(this.signaturePad.isEmpty()) return;
+            const data =  await (await fetch(this.signaturePad.toDataURL())).blob();
+            await $wire.upload('signature.file',data, this.upload.bind(this));
+        },
+        // TODO: Add resizeCanvas method to resize listener
+        resizeCanvas() {
```
[the remark this answers is not in the document]
Comment (sajberkurajber90) [reply resources/js/components/signature-pad.js:62]: it's todo

File: resources/views/livewire/order/public-link.blade.php (hunk 6 above)
Comment (sajberkurajber90) [reply resources/views/livewire/order/public-link.blade.php:19]: if both are hidden(x-show) - same position of message can not be kept when one of them is displayed

File: resources/views/livewire/order/public-link.blade.php (hunk 7 above)
Comment (sajberkurajber90) [reply resources/views/livewire/order/public-link.blade.php:12]: if both are hidden(x-show) - same position of message can not be kept when one of them is displayed

File: src/Livewire/Order/PublicLink.php [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                exception_to_notifications($e, $this);
+
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    public function downloadSignatureAsUrlData(Media $media): string|bool
```
[the remark this answers is not in the document]
Comment (sajberkurajber90) [reply src/Livewire/Order/PublicLink.php:59]: is used in signature-pad.js to load signature if already submited

  async init() {
            // init signature pad
            this.signaturePad = new SignaturePad($refs.canvas, {backgroundColor: 'rgba(255, 255, 255, 1)'});
            // if signature is already saved - just display it on canvas but don't allow to draw
            if ($wire.signature.stagedFiles.length > 0) {
                this.id = $wire.signature.id;
                await this.signaturePad.fromDataURL(await $wire.downloadSignatureAsUrlData(this.id));
                this.signaturePad.off();

                return;
            }
            // if signature is not saved - allow to draw and clear after first stroke
            this.signaturePad.addEventListener('afterUpdateStroke', this.strokeHandler.bind(this));
        },

File: resources/views/layouts/empty.blade.php (hunk 8 above)
Comment (sajberkurajber90) [reply resources/views/layouts/empty.blade.php:1]: livewire loaded twice using this approach

File: resources/views/livewire/order/public-link.blade.php [hunk 12]
```diff
@@ -0,0 +1,54 @@
+<div>
+
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:2]: remove empty line

File: resources/views/livewire/order/public-link.blade.php [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+@if(!is_null($className))
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:3]: indent 1 tab

File: resources/views/livewire/order/public-link.blade.php [hunk 14]
```diff
@@ -0,0 +1,54 @@
+<div>
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:1]: add margin, too close to the end of the order

File: resources/views/livewire/public-link-generator.blade.php [hunk 15]
```diff
@@ -0,0 +1,53 @@
+<div>
+    @if(!empty($signed_urls))
+        <div class="mb-4 flex flex-col">
+            <p>{{__('Signed Documents')}}:</p>
+            @foreach($signed_urls as $url)
+                <p class="mt-2">{{__($url)}}</p>
+            @endforeach
+        </div>
+    @endif
+    @if(!empty($unsigned_documents))
```
Comment (patrickweh) [inline resources/views/livewire/public-link-generator.blade.php:10]: dont use empty, just ! $unsignedDocuments
also change to camelCase

File: resources/views/livewire/public-link-generator.blade.php [hunk 16]
```diff
@@ -0,0 +1,53 @@
+<div>
+    @if(!empty($signed_urls))
```
Comment (patrickweh) [inline resources/views/livewire/public-link-generator.blade.php:2]: dont use empty, just ! $signedUrls
also change to camelCase

File: resources/views/livewire/public-link-generator.blade.php [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+                @foreach($unsigned_documents as $doc)
+                    <x-dropdown.item
+                        wire:click="setPublicLink('{{$doc}}')">
+                        {{ __($doc) }}
+                    </x-dropdown.item>
+                @endforeach
+            </x-dropdown>
+        </div>
+    @endif
+    @if(!empty($generated_urls))
```
Comment (patrickweh) [inline resources/views/livewire/public-link-generator.blade.php:27]: dont use empty, just ! $generatedUrls
also change to camelCase

File: resources/views/livewire/public-link-generator.blade.php [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        <div class="mt-4">
+            @foreach($generated_urls as $label => $link)
+                <x-input
+                    class="mb-2"
+                    label="{{__($label)}}"
```
Comment (patrickweh) [inline resources/views/livewire/public-link-generator.blade.php:32]: :label="__($label)"

File: resources/views/livewire/public-link-generator.blade.php [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+                    readonly
+                    value="{{$link}}"
+                    type="text"
+                    x-ref="link"
```
Comment (patrickweh) [inline resources/views/livewire/public-link-generator.blade.php:36]: are you sure this works with the same x-ref inside a blade foreach?

File: routes/frontend/web.php [hunk 20]
```diff
@@ -208,4 +209,6 @@
                 return $media;
             })->name('media.private');
         });
+
+        Route::get('/order-public/{order:uuid}', PublicLink::class)->name('order.public');
```
Comment (patrickweh) [inline routes/frontend/web.php:213]: as this should be protected by the signature you should move this to the block where the signed middleware is used

File: src/Livewire/Order/PublicLink.php [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+    public MediaForm $signature;
+
+    public ?string $className;
+
+    #[Url]
+    public ?string $orderType = null;
+
+    public function mount(): void
+    {
+
```
Comment (patrickweh) [inline src/Livewire/Order/PublicLink.php:32]: empty line

File: src/Livewire/Order/PublicLink.php [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+                ->whereJsonContains('custom_properties->order_type', $this->orderType)
+                ->first();
+
+            $this->signature->fill($media ?? []);
+        } // false
+    }
+
+    public function save(): bool
+    {
+
```
Comment (patrickweh) [inline src/Livewire/Order/PublicLink.php:50]: empty line

File: src/Livewire/PublicLinkGenerator.php [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+
+    public array $signed_urls = [];
+
+    public array $unsigned_documents = [];
+
+    public array $generated_urls = [];
+
+    public function mount(): void
+    {
+        $this->signed_urls = array_map(function (array $item) {
```
Comment (patrickweh) [inline src/Livewire/PublicLinkGenerator.php:24]: crazy formatting here…

File: src/Livewire/PublicLinkGenerator.php [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+use Livewire\Component;
+
+class PublicLinkGenerator extends Component
+{
+    #[Modelable]
+    public OrderForm $order;
+
+    public array $signed_urls = [];
+
+    public array $unsigned_documents = [];
```
Comment (patrickweh) [inline src/Livewire/PublicLinkGenerator.php:18]: change to camelCase

File: src/Livewire/PublicLinkGenerator.php [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+use Illuminate\Support\Facades\URL;
+use Livewire\Attributes\Modelable;
+use Livewire\Component;
+
+class PublicLinkGenerator extends Component
+{
+    #[Modelable]
+    public OrderForm $order;
+
+    public array $signed_urls = [];
```
Comment (patrickweh) [inline src/Livewire/PublicLinkGenerator.php:16]: change to camelCase

File: src/Livewire/PublicLinkGenerator.php [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+
+    public array $unsigned_documents = [];
+
+    public array $generated_urls = [];
```
Comment (patrickweh) [inline src/Livewire/PublicLinkGenerator.php:20]: change to camelCase

File: src/Models/Order.php [hunk 27]
```diff
@@ -45,11 +45,10 @@
 use TeamNiftyGmbH\DataTable\Casts\Money;
 use TeamNiftyGmbH\DataTable\Casts\Percentage;
 use TeamNiftyGmbH\DataTable\Contracts\InteractsWithDataTables;
-use TeamNiftyGmbH\DataTable\Traits\BroadcastsEvents;
 
 class Order extends Model implements HasMedia, InteractsWithDataTables, OffersPrinting
 {
-    use BroadcastsEvents, Commentable, Communicatable, Filterable, HasAdditionalColumns, HasClientAssignment,
```
Comment (patrickweh) [inline src/Models/Order.php:52]: why did you remove the broadcasting?

**Review by patrickweh — CHANGES_REQUESTED** [review]

File: resources/views/livewire/order/public-link.blade.php [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+                    </div>
+                    <canvas x-ref="canvas" width=500 height=200 class="rounded-md h-auto">
+                    </canvas>
+                </div>
+                <div class="mt-4 h-8 mb-4 flex justify-end w-full space-x-5">
+                    <x-button
+                        x-cloak
+                        x-show="!isEmpty"
+                        x-on:click="clear"
+                        class="bg-red-500 rounded-md px-4 py-2 text-center font-medium text-white shadow-sm ring-1 ring-slate-700/10 hover:bg-red-600"
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:35]: remove the whole class, just add negative

File: resources/views/livewire/order/public-link.blade.php [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+                        :label="__('Clear')"
+                    />
+                    <x-button
+                        x-cloak
+                        x-show="!isEmpty"
+                        x-on:click="save"
+                        class="bg-primary-600 rounded-md px-4 py-2 text-center font-medium text-white shadow-sm ring-1 ring-slate-700/10"
```
Comment (patrickweh) [inline resources/views/livewire/order/public-link.blade.php:42]: remove the whole class, just add primary

**Review by patrickweh — CHANGES_REQUESTED** [review]

File: src/Livewire/PublicLinkGenerator.php (hunk 23 above)
Comment (sajberkurajber90) [reply src/Livewire/PublicLinkGenerator.php:24]: should i change somethig ?

File: resources/views/livewire/public-link-generator.blade.php (hunk 18 above)
Comment (sajberkurajber90) [reply resources/views/livewire/public-link-generator.blade.php:32]: livewire i spits fire at me if i use that - i think :label takes the variable not string as a parameter

File: resources/views/livewire/public-link-generator.blade.php (hunk 19 above)
Comment (sajberkurajber90) [reply resources/views/livewire/public-link-generator.blade.php:36]: you are right - it worked last time - now not - i fix

File: src/Models/Order.php (hunk 27 above)
Comment (sajberkurajber90) [reply src/Models/Order.php:52]: i was probably high on something - i will revert - tnx for notice

**Review by SirSplasch — CHANGES_REQUESTED** [review]

**Review by SirSplasch — APPROVED** [review]
